#  LD score regression on auotoimmunity GWAS
##  in this script the selected summary stats will be used to run ldsc, munge and sum stats function
#the have to be in the same order in the function

#The tutorial and info on the package and how to run the code are here:  
#https://github.com/GenomicSEM/GenomicSEM/wiki/5.-User-Specified-Models-with-SNP-Effects

#This script, step4, is for preparing the sumstats to run the GWAS by using the two factor model
#specified in step3. 
# 'Please also note that the sumstats function requires that the variables be listed -
#in the same order as they are listed for the ldsc function.'

#This means that I have to run again the analysis from the munge step for only 
#the GWAS we are keeping in the model. 
#The GWAS included in the two factor model are croh, uc, psc, jia, sle, ra.
#as in step3


#--- Load the libraries--------------------------------------
library(data.table)
library(GenomicSEM)
library(tidyr)
library(dplyr)
library(corrplot)

#----CHeck the standard errors--------------------------------------------------

#----- Function to check if the SE are STD error of log(OD)-----
is_se_logB <- function(BETA,SE, PVALUE) {
  p_calculated <- 2*pnorm((abs(BETA) / SE),lower.tail = F)
  p_reported <- PVALUE
  data.frame(p_calculated, p_reported)
}


#---- crohn SE -----------------------------------------------------------------
cd <- fread('outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/crohn_delange-2017.txt', data.table = F)
head(cd) 
is_se_logB(cd$effect, cd$StdErr, cd$p) #SE logistic beta (effect column is beta as there are negative values)

#----- uc SE --------------------------------------------------------------------
uc <- fread('outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/uc_delange-2017.txt', data.table = F)
head(uc)
is_se_logB(uc$effect, uc$StdErr, uc$p) #SE logistic beta (effect column is beta as there are negative values)

#----- psc SE -------------------------------------------------------------------
psc <- fread('outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/psc_ji-2016.txt', data.table = F)
head(psc)

is_se_logB(log(psc$Effect), psc$SE, psc$P) #SE of logistic beta

#----- jia SE ------------------------------------------------------------------

jia  <- fread('outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/jia_beta_lopezisac-2020.txt', data.table = F) 
head(jia)
is_se_logB(jia$effect, jia$SE, jia$p) #SE of logististi bete

#----- sle SE ------------------------------------------------------------------

sle <- fread('outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/sle_beta_bentham-2015.txt', data.table = F)
head(sle)
is_se_logB(BETA = sle$effect, SE = sle$se, PVALUE = sle$p) #SE of logistic beta (effect column is beta as there are negative values)

#----- ra eu okada SE -------------------------------------------------------------------

ra <- fread('outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/ra_eu_okada-2014.txt', data.table = F)
head(ra)
is_se_logB(BETA = ra$effect, SE = ra$SE, PVALUE = ra$p) #SE of logistic beta

#-----t1d SE----------------------------------------------------------------------

t1d <- fread('outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/t1d_chiou-2021.txt', data.table = F)
a <- is_se_logB(BETA = t1d$effect, SE = t1d$SE, PVALUE = t1d$p) #SE of logistic BETA


#------asthma 2 SE -------------------------------------------------------------

asthma <- fread('outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/asthma_demeanis-2018.txt', data.table = F)
head(asthma)
is_se_logB(asthma$effect, asthma$SE, asthma$p) #SE of logistic BETA, the analysis is logfistic regression

#----derma SE ---------------------------------------------------------------
derma <- fread('outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/derma_sliz-2021.txt', data.table = F)
head(derma)
is_se_logB(derma$effect, derma$SE, derma$p) #SE of logistic beta, logistic regression checked on paper


#---- Munge the summary stats --------------------------------------------------

#Function for sample prevalence calculation-------------------------------------
#a function for doing it rapidly from csv files 
#csv files generated by me, from info in the papers
calculate_prevalence <- function(path_of_csv){
  #read the file and remove the NA columns and rows
  preva_csv <- read.csv(file = path_of_csv , header = T, sep = ';')
  preva_csv <-  preva_csv[, c('cases', 'controls')]
  preva_csv_noNA <- preva_csv[complete.cases(preva_csv), ]
  
  #calculated effective prevalence 
  #calculate sample prevalence for each cohort
  preva_csv_noNA$v <-preva_csv_noNA$cases/(preva_csv_noNA$cases+preva_csv_noNA$controls)
  #calculate cohort specific effective sample size
  preva_csv_noNA$EffN<-4*preva_csv_noNA$v*(1-preva_csv_noNA$v)*(preva_csv_noNA$cases+preva_csv_noNA$controls)
  #calculate sum of effective sample size: 
  eff_sample_size <- round(sum(preva_csv_noNA$EffN), 3)
  
  #print number of cases and controls, and effective sample size
  cat(paste0( 'Number of cases  ' , sum(preva_csv_noNA$cases), '\n', 
              'Number of contros  ', sum(preva_csv_noNA$controls), '\n', 
              'Effective sample size  ', round(sum(preva_csv_noNA$EffN), 3)
  ))
  
  #output the prevalenc when assigned to a variable 
  invisible(round(eff_sample_size, 3))
}

#---- munge function -----------------------------------------------------------

file_names <- c('outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/t1d_chiou-2021.txt', 
                'outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/crohn_delange-2017.txt', 
                'outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/uc_delange-2017.txt',
                'outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/psc_ji-2016.txt',
                'outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/jia_beta_lopezisac-2020.txt',
                'outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/sle_beta_bentham-2015.txt',
                'outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/ra_eu_okada-2014.txt',
                'outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/asthma_demeanis-2018.txt',
                'outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/derma_sliz-2021.txt'
                )
traits <- c('t1d','crohn', 'uc', 'psc', 'jia', 'sle', 'ra',  'asthma', 'derma')

t1d_p <- calculate_prevalence('Prevalences/CSV_prevalences/t1d_chiou-2021.csv')
crohn_p <- calculate_prevalence('Prevalences/CSV_prevalences/crohn_delange-2017.csv')
uc_p <- calculate_prevalence('Prevalences/CSV_prevalences/uc_delange-2017.csv')
psc_p <- 14890
jia_p <- 12501
sle_p <- calculate_prevalence('Prevalences/CSV_prevalences/sle_benthman-2015.csv')
ra_p <- calculate_prevalence('Prevalences/CSV_prevalences/ra_okada-2014_only-eu.csv')
asthma_p <- calculate_prevalence('Prevalences/CSV_prevalences/asthma_demeanis-2018.csv')
derma_p <- calculate_prevalence('Prevalences/CSV_prevalences/derma_sliz-2021.csv')

munge(files = file_names, hm3 = 'SNP/w_hm3.snplist', 
      N = c(t1d_p, crohn_p, uc_p, psc_p, jia_p, sle_p, ra_p, asthma_p, derma_p), 
      trait.names = traits)



#-----ldsc function ------------------------------------------------------------
munged_files <- c('outputs/version3/04_output_sumstats-function/GWAS-10-05-2022/munge_output/t1d.sumstats.gz',
                  'outputs/version3/04_output_sumstats-function/GWAS-10-05-2022/munge_output/crohn.sumstats.gz',
                  'outputs/version3/04_output_sumstats-function/GWAS-10-05-2022/munge_output/uc.sumstats.gz',
                  'outputs/version3/04_output_sumstats-function/GWAS-10-05-2022/munge_output/psc.sumstats.gz',
                  'outputs/version3/04_output_sumstats-function/GWAS-10-05-2022/munge_output/jia.sumstats.gz', 
                  'outputs/version3/04_output_sumstats-function/GWAS-10-05-2022/munge_output/sle.sumstats.gz', 
                  'outputs/version3/04_output_sumstats-function/GWAS-10-05-2022/munge_output/ra.sumstats.gz', 
                  'outputs/version3/04_output_sumstats-function/GWAS-10-05-2022/munge_output/asthma.sumstats.gz', 
                  'outputs/version3/04_output_sumstats-function/GWAS-10-05-2022/munge_output/derma.sumstats.gz')

names = c('t1d','crohn', 'uc', 'psc', 'jia', 'sle', 'ra',  'asthma', 'derma')




ldsc_gwas_10052022 <- ldsc(traits = munged_files, 
                      sample.prev = c(.5, .5, .5, 0.1928, 0.2643, .5, .5, .5, .5),  
                     population.prev = c(0.095, 0.001, 0.0003, 0.000050, 0.000447, 0.000500,  0.004600, 0.035700, 0.20 ), 
                     trait.names = names,
                     ld = "ldscores/eur_w_ld_chr",
                     wld= "ldscores/eur_w_ld_chr", stand = T)

saveRDS(ldsc_gwas_10052022, 'outputs/version3/04_output_sumstats-function/GWAS-10-05-2022/ldsc_output/ldsc_gwas_10052022.RDS')
ldsc_ok <- readRDS('outputs/version3/04_output_sumstats-function/GWAS-10-05-2022/ldsc_output/ldsc_gwas_10052022.RDS')
rownames(ldsc_ok$S_Stand) <- colnames(ldsc_ok$S_Stand)
corrplot(ldsc_ok$S_Stand, order = 'hclust', addCoef.col = 'black', type = 'upper', is.corr = T)


#---- Two factor model ---------------------------------------------------------
aid_model <-'F1 =~ NA*crohn + uc  + psc  
F2 =~ NA*jia + sle + ra + t1d 
F3 =~ NA*asthma + derma 

F1~~F2
F1~~F3
F2~~F3
F1 ~~ 1*F1
F2 ~~ 1*F2
F3 ~~ 1*F3
derma~~a*derma
a>0.001
'


#run the model
aid_factor <-usermodel(ldsc_ok, estimation = "DWLS", model = aid_model, CFIcalc = TRUE, std.lv = TRUE, imp_cov = FALSE)
aid_factor



#----Sumstat function-----------------------------------------------------------


file_names_ok <- c('outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/t1d_chiou-2021.txt', 
                'outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/crohn_delange-2017.txt', 
                'outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/uc_delange-2017.txt',
                'outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/psc_ji-2016.txt',
                'outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/jia_beta_lopezisac-2020.txt',
                'outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/sle_beta_bentham-2015.txt',
                'outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/ra_eu_okada-2014.txt',
                'outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/asthma_demeanis-2018.txt',
                'outputs/version3/01_output_prepare-sumstats/Sumstats_ready_for_munge/derma_sliz-2021.txt'
)



t1d_p <- calculate_prevalence('Prevalences/CSV_prevalences/t1d_chiou-2021.csv')
crohn_p <- calculate_prevalence('Prevalences/CSV_prevalences/crohn_delange-2017.csv')
uc_p <- calculate_prevalence('Prevalences/CSV_prevalences/uc_delange-2017.csv')
psc_p <- 14890
jia_p <- 12501
sle_p <- calculate_prevalence('Prevalences/CSV_prevalences/sle_benthman-2015.csv')
ra_p <- calculate_prevalence('Prevalences/CSV_prevalences/ra_okada-2014_only-eu.csv')
asthma_p <- calculate_prevalence('Prevalences/CSV_prevalences/asthma_demeanis-2018.csv')
derma_p <- calculate_prevalence('Prevalences/CSV_prevalences/derma_sliz-2021.csv')




aid_sumstats_2F_noRA <-sumstats(files = file_names_ok, ref = 'SNP/reference.1000G.maf.0.005.txt.gz',
                                trait.names = c('t1d','crohn', 'uc', 'psc', 'jia', 'sle', 'ra',  'asthma', 'derma') ,
                                se.logit =  c(T,T,T,T,T,T,T,T,T),  
                                OLS = c(F,F,F,F,F,F,F,F,F), 
                                linprob = c(F,F,F,F,F,F,F,F,F), 
                                N = c(t1d_p, crohn_p, uc_p, psc_p, jia_p, sle_p, ra_p, asthma_p, derma_p), 
                                parallel = T,
                                keep.indel= F 
)

saveRDS(aid_sumstats_2F_noRA, file= 'outputs/version3/04_output_sumstats-function/GWAS-10-05-2022/sumstats_output/aid_sumstats_100522.RDS')

#aid_sumstats_noPBC <- readRDS('outputs/version3/04_output_sumstats-function/aid_sumstats_newRA.RDS') 

#----------------------------------------------------------------------------



















