#  LD score regression on auotoimmunity GWAS
## Version3  will be a replication of the version 2 that might be wrong in the 
## allele orientation 

#The tutorial and info on the package and how to run the code are here:  
#https://github.com/GenomicSEM/GenomicSEM/wiki/5.-User-Specified-Models-with-SNP-Effects

#This script, step4, is for preparing the sumstats to run the GWAS by using the two factor model
#specified in step3. 
# 'Please also note that the sumstats function requires that the variables be listed -
#in the same order as they are listed for the ldsc function.'

#This means that I have to run again the analysis from the munge step for only 
#the GWAS we are keeping in the model. 
#The GWAS included in the two factor model are croh, uc, psc, jia, pbc, sle, ra.
#as in V3_step3


#--- Load the libraries--------------------------------------
library(data.table)
library(GenomicSEM)
library(tidyr)
library(dplyr)
library(corrplot)


#---- Munge the summary stats from the already prepared GWAS as un V3_step1-----

#Function for sample prevalence calculation-------------------------------------
#a function for doing it rapidly from csv files 
#csv files generated by me, from info in the papers
calculate_prevalence <- function(path_of_csv){
  #read the file and remove the NA columns and rows
  preva_csv <- read.csv(file = path_of_csv , header = T, sep = ';')
  preva_csv <-  preva_csv[, c('cases', 'controls')]
  preva_csv_noNA <- preva_csv[complete.cases(preva_csv), ]
  
  #calculated effective prevalence 
  #calculate sample prevalence for each cohort
  preva_csv_noNA$v <-preva_csv_noNA$cases/(preva_csv_noNA$cases+preva_csv_noNA$controls)
  #calculate cohort specific effective sample size
  preva_csv_noNA$EffN<-4*preva_csv_noNA$v*(1-preva_csv_noNA$v)*(preva_csv_noNA$cases+preva_csv_noNA$controls)
  #calculate sum of effective sample size: 
  eff_sample_size <- round(sum(preva_csv_noNA$EffN), 3)
  
  #print number of cases and controls, and effective sample size
  cat(paste0( 'Number of cases  ' , sum(preva_csv_noNA$cases), '\n', 
              'Number of contros  ', sum(preva_csv_noNA$controls), '\n', 
              'Effective sample size  ', round(sum(preva_csv_noNA$EffN), 3)
  ))
  
  #output the prevalenc when assigned to a variable 
  invisible(round(eff_sample_size, 3))
}

#---- munge function -----------------------------------------------------------

file_names <- c('Outputs/Version3/Sumstats_ready_for_munge/crohn_delange-2017.txt',
                'Outputs/Version3/Sumstats_ready_for_munge/uc_delange-2017.txt', 
                'Outputs/Version3/Sumstats_ready_for_munge/psc_ji-2016.txt',
                'Outputs/Version3/Sumstats_ready_for_munge/jia_lopezisac-2020.txt',
                'Outputs/Version3/Sumstats_ready_for_munge/pbc_cordell-2015.txt',
                'Outputs/Version3/Sumstats_ready_for_munge/sle_beta_bentham-2015.txt',
                'Outputs/Version3/Sumstats_ready_for_munge/ra_okada-2014.txt') 



crohn_p <- calculate_prevalence('Prevalences/CSV_prevalences/crohn_delange-2017.csv')
uc_p <- calculate_prevalence('Prevalences/CSV_prevalences/uc_delange-2017.csv')
psc_p <- 2871 + 12019
jia_p <- 3305 + 9196 
pbc_p <- calculate_prevalence('Prevalences/CSV_prevalences/pbc_cordell-2015.csv')
sle_p <- calculate_prevalence('Prevalences/CSV_prevalences/sle_benthman-2015.csv')
ra_p <- calculate_prevalence('Prevalences/CSV_prevalences/ra_okada-2014.csv')

prevalences <- c(crohn_p, uc_p, psc_p, jia_p, pbc_p, sle_p, ra_p)
trait.names = c('croh', 'uc', 'psc', 'jia', 'pbc', 'sle', 'ra') 
hm3 = 'SNP/w_hm3.snplist'

munge(files = file_names, hm3 = 'SNP/w_hm3.snplist', N = prevalences, trait.names = trait.names )

#----LDSC-----------------------------------------------------------------------

# Load the table that contains all the info on the GWAS
GWAS_info <- readRDS('Outputs/Version3/GWAS_info_table')
#keep only the one I want ot use here
GWAS_info_2 <- GWAS_info[ c('croh', 'uc', 'psc', 'jia', 'pbc', 'sle', 'ra') , ]

#-----ldsc function ------------------------------------------------------------
munged_files <- c('Outputs/Version3/Step4/munged/croh.sumstats.gz', 
                  'Outputs/Version3/Step4/munged/uc.sumstats.gz', 
                  'Outputs/Version3/Step4/munged/psc.sumstats.gz',
                  'Outputs/Version3/Step4/munged/jia.sumstats.gz', 
                  'Outputs/Version3/Step4/munged/pbc.sumstats.gz', 
                  'Outputs/Version3/Step4/munged/sle.sumstats.gz', 
                  'Outputs/Version3/Step4/munged/ra.sumstats.gz')

names = c('croh', 'uc', 'psc', 'jia', 'pbc', 'sle', 'ra') 

ldsc_step4 <- ldsc(traits = munged_files, sample.prev = GWAS_info_2$sample.prev, 
     population.prev = GWAS_info_2$population.prev, trait.names = names,
     ld = "ldscores/eur_w_ld_chr",
     wld= "ldscores/eur_w_ld_chr", stand = T)
saveRDS(ldsc_step4, 'Outputs/Version3/Step4/ldsc_V3_step4')
#ldsc4_step4 <- readRDS('Outputs/Version3/Step4/ldsc_V3_step4')
rownames(ldsc_step4$S_Stand) <- colnames(ldsc_step4$S_Stand)
corrplot(ldsc_step4$S_Stand, order = 'hclust', addCoef.col = 'black', type = 'upper')


#---- Two factor model ---------------------------------------------------------
aid_model <-'F1 =~ NA*croh + uc  + psc 
             F2 =~ NA*jia + pbc + sle + ra
F1~~F2'

#run the model
aid_factor <-usermodel(ldsc_step4, estimation = "DWLS", model = aid_model, CFIcalc = TRUE, std.lv = TRUE, imp_cov = FALSE)

#print the two factor model results
saveRDS(aid_factor, file = 'Outputs/Version3/Step4/aid_twofactor') 
aid_factor <- readRDS('Outputs/Version3/Step4/aid_twofactor')

#----Sumstats function----------------------------------------------------------

#----SE of logistic BETA--------------------------------------------------------

#----- Function to check if the SE are STD error of log(OD)-----
is_se_logB <- function(BETA,SE, PVALUE) {
    p_calculated <- 2*pnorm((abs(BETA) / SE),lower.tail = F)
    p_reported <- PVALUE
    data.frame(p_calculated, p_reported)
}

#------- Function to compute SE from CI or from Betas-----

add_SE <- function(OR, PVALUE, upper_CI, use_chi = F, add_se=F, sum_stat) {
  
  #calculate SE from Confidence intervals 
  if(use_chi == F){
      se_column <- (log(upper_CI) - log(OR) )/qnorm(0.975)
      p_calculated <- 2*pnorm((abs(log(OR)) / se_column),lower.tail = F)
      p_reported <- PVALUE
    
      #add SE column
      if(add_se == T) {
        sum_stat2 <- mutate(sum_stat, SE = (log(upper_CI) - log(OR) )/qnorm(0.975) )
        return(sum_stat2) }
    
    
    #calculate SE from X2 distribution
  } else {
      chi <-  qchisq(1-PVALUE, df=1)
      se_column <- sqrt((abs(log(OR)))^2/chi)
      p_calculated <- 2*pnorm((abs(log(OR)) / se_column),lower.tail = F)
      p_reported <- PVALUE
    
      #add SE column
      if(add_se == T) {
        sum_stat2 <- mutate(sum_stat, SE = sqrt((abs(log(OR)))^2/chi) )
        return(sum_stat2) }
  }
}
   
  


#---- crohn SE -----------------------------------------------------------------
cd <- fread('Outputs/Version3/Sumstats_ready_for_munge/crohn_delange-2017.txt', data.table = F)
head(cd)
is_se_logB(cd$effect, cd$StdErr, cd$p) #SE log(OD)

#----- uc SE --------------------------------------------------------------------
uc <- fread('Outputs/Version3/Sumstats_ready_for_munge/uc_delange-2017.txt', data.table = F)
head(uc)
is_se_logB(uc$effect, uc$StdErr, uc$p) #SE log(OD)

#----- psc SE -------------------------------------------------------------------
psc <- fread('Outputs/Version3/Sumstats_ready_for_munge/psc_ji-2016.txt', data.table = F)
head(psc)

is_se_logB(log(psc$Effect), psc$SE, psc$P) #SE of Odds Ratio

#----- jia SE ------------------------------------------------------------------

jia  <- fread('Outputs/Version3/Sumstats_ready_for_munge/jia_lopezisac-2020.txt', data.table = F) 
head(jia)

#there are a bunch of columns we do not care and the SE have to be calculated from the CI
jia_ok <- jia %>% select(-c(frequentist_add_beta_1, frequentist_add_se_1, alternate_ids ))

#here derive the SE from the CHI square to avoid rounding issues. 
#the effect seems to be an Odds Ratio
jia_ok <- add_SE(OR = jia_ok$effect, PVALUE = jia_ok$p, upper_CI = jia_ok$all_OR_upper, 
       use_chi =  T, add_se = T, sum_stat = jia_ok )

head(jia_ok)

is_se_logB(BETA = log(jia_ok$effect), SE = jia_ok$SE, PVALUE = jia_ok$p) #now is a SE of logistic Beta

#save the file with SE
fwrite(jia_ok, file='Outputs/Version3/Sumstats_ready_for_munge/jia_lopezisac-2020_SE_.txt', 
       sep = '\t', col.names = T, row.names = F, quote = F)
#----


















