#  LD score regression on auotoimmunity GWAS
## Version3  will be a replication of the version 2 that might be wrong in the 
## allele orientation 

#The tutorial and info on the package and how to run the code are here:  
#https://github.com/GenomicSEM/GenomicSEM/wiki/5.-User-Specified-Models-with-SNP-Effects

#This script, step4, is for preparing the sumstats to run the GWAS by using the two factor model
#specified in step3. 

#--- Load the libraries--------------------------------------
library(data.table)
library(GenomicSEM)
library(tidyr)
library(dplyr)

#-----Important comment --------------------------------------------------------

# 'Please also note that the sumstats function requires that the variables be listed -
#in the same order as they are listed for the ldsc function.'

#This means that I have to run again the analysis from the munge step for only 
#the GWAS we are keeping in the model. 
#The GWAS included in the two factor model are croh, uc, psc, jia, pbc, sle, ra.
#as in V3_step3

#------End of comment-----------------------------------------------------------

#----Munge ---------------------------------------------------------------------

#Function for sample prevalence calculation-------------------------------------
#a function for doing it rapidly from csv files 
#csv files generated by me, from info in the papers

calculate_prevalence <- function(path_of_csv){
  #read the file and remove the NA columns and rows
  preva_csv <- read.csv(file = path_of_csv , header = T, sep = ';')
  preva_csv <-  preva_csv[, c('cases', 'controls')]
  preva_csv_noNA <- preva_csv[complete.cases(preva_csv), ]
  
  #calculated effective prevalence 
  #calculate sample prevalence for each cohort
  preva_csv_noNA$v <-preva_csv_noNA$cases/(preva_csv_noNA$cases+preva_csv_noNA$controls)
  #calculate cohort specific effective sample size
  preva_csv_noNA$EffN<-4*preva_csv_noNA$v*(1-preva_csv_noNA$v)*(preva_csv_noNA$cases+preva_csv_noNA$controls)
  #calculate sum of effective sample size: 
  eff_sample_size <- round(sum(preva_csv_noNA$EffN), 3)
  
  #print number of cases and controls, and effective sample size
  cat(paste0( 'Number of cases  ' , sum(preva_csv_noNA$cases), '\n', 
              'Number of contros  ', sum(preva_csv_noNA$controls), '\n', 
              'Effective sample size  ', round(sum(preva_csv_noNA$EffN), 3)
  ))
  
  #output the prevalenc when assigned to a variable 
  invisible(round(eff_sample_size, 3))
}


#---- Munge the summary stats from the already prepared GWAS as un V3_step1-----

file_names <- c('Outputs/Version3/Sumstats_ready_for_munge/crohn_delange-2017.txt',
                'Outputs/Version3/Sumstats_ready_for_munge/uc_delange-2017.txt', 
                'Outputs/Version3/Sumstats_ready_for_munge/psc_ji-2016.txt',
                'Outputs/Version3/Sumstats_ready_for_munge/jia_lopezisac-2020.txt',
                'Outputs/Version3/Sumstats_ready_for_munge/pbc_cordell-2015.txt',
                'Outputs/Version3/Sumstats_ready_for_munge/sle_bentham-2015.txt',
                'Outputs/Version3/Sumstats_ready_for_munge/ra_okada-2014.txt') 



crohn_p <- calculate_prevalence('Prevalences/CSV_prevalences/crohn_delange-2017.csv')
uc_p <- calculate_prevalence('Prevalences/CSV_prevalences/uc_delange-2017.csv')
psc_p <- 2871 + 12019
jia_p <- 3305 + 9196 
pbc_p <- calculate_prevalence('Prevalences/CSV_prevalences/pbc_cordell-2015.csv')
sle_p <- calculate_prevalence('Prevalences/CSV_prevalences/sle_benthman-2015.csv')
ra_p <- calculate_prevalence('Prevalences/CSV_prevalences/ra_okada-2014.csv')

prevalences <- c(crohn_p, uc_p, psc_p, jia_p, pbc_p, sle_p, ra_p)
trait.names = c('croh', 'uc', 'psc', 'jia', 'pbc', 'sle', 'ra') 
hm3 = 'SNP/w_hm3.snplist'

munge(files = file_names, hm3 = 'SNP/w_hm3.snplist', N = prevalences, trait.names = trait.names )

#----LDSC-----------------------------------------------------------------------

# Crete a table that contains all the info on the GWAS--------------------------

traits <- c('Outputs/Version3/Munged-Sumstats/allergies.sumstats.gz',  
            'Outputs/Version3/Munged-Sumstats/ms_2.sumstats.gz' ,
            'Outputs/Version3/Munged-Sumstats/alzheimer_1.sumstats.gz',
            'Outputs/Version3/Munged-Sumstats/alzheimer_2.sumstats.gz',
            'Outputs/Version3/Munged-Sumstats/armfat.sumstats.gz',
            'Outputs/Version3/Munged-Sumstats/asthma_1.sumstats.gz',
            'Outputs/Version3/Munged-Sumstats/asthma_2.sumstats.gz',
            
            'Outputs/Version3/Munged-Sumstats/celiac.sumstats.gz',
            'Outputs/Version3/Munged-Sumstats/crohn.sumstats.gz',
            'Outputs/Version3/Munged-Sumstats/jia.sumstats.gz',
            'Outputs/Version3/Munged-Sumstats/ms_1.sumstats.gz',
            'Outputs/Version3/Munged-Sumstats/pbc.sumstats.gz',
            
            'Outputs/Version3/Munged-Sumstats/psc.sumstats.gz', 
            'Outputs/Version3/Munged-Sumstats/ra.sumstats.gz',
            'Outputs/Version3/Munged-Sumstats/sle.sumstats.gz', 
            'Outputs/Version3/Munged-Sumstats/thyro.sumstats.gz',
            'Outputs/Version3/Munged-Sumstats/uc.sumstats.gz'
) 

trait.names <- c( 'allergies','ms_2','alzheimer_1', 'alzheimer_2', 'armfat', 'asthma_1', 'asthma_2', 
                  'celiac', 'crohn', 'jia', 'ms_1', 'pbc', 
                  'psc', 'ra', 'sle', 'thyro', 'uc')
sample.prev <-  c(.5, .5, .5, .5, NA, .5, (64538/(64538 + 239321)),
                  .5, .5, (3305/(3305 + 9196)), (9772 /(9772 + 16849)), .5, 
                  ( 2871 /(2871 + 12019)), .5, .5, .5, .5 )

population.prev <-  c((0.20 ),(35.9/100000), (0.058), (0.058), (NA), (0.0357), (0.0357), 
                      (0.014), (100/100000), (44.7/100000), (35.9/100000), (10/100000),
                      (5/100000), (460/100000), (50/100000), (0.05), (30/100000) )

GWAS_info <- data.frame(trait.names, traits, sample.prev, population.prev)
row.names(GWAS_info) <- GWAS_info$trait.names


#keep only the one I want ot use here
GWAS_info_2 <- GWAS_info[ c('croh', 'uc', 'psc', 'jia', 'pbc', 'sle', 'ra') , ]

#---- function ------------------------------------------------------------
munged_files <- c('Outputs/Version3/Munged-Sumstats_step4/croh.sumstats.gz', 
                  'Outputs/Version3/Munged-Sumstats_step4/uc.sumstats.gz', 
                  'Outputs/Version3/Munged-Sumstats_step4/psc.sumstats.gz',
                  'Outputs/Version3/Munged-Sumstats_step4/jia.sumstats.gz', 
                  'Outputs/Version3/Munged-Sumstats_step4/pbc.sumstats.gz', 
                  'Outputs/Version3/Munged-Sumstats_step4/sle.sumstats.gz', 
                  'Outputs/Version3/Munged-Sumstats_step4/ra.sumstats.gz')

trait.names = c('croh', 'uc', 'psc', 'jia', 'pbc', 'sle', 'ra') 
population_prevalence <-   
  
ld <- "ldscores/eur_w_ld_chr"
wld <- "ldscores/eur_w_ld_chr"




#---- Function for standard error of logistic beta------------------------------

#The package needs to know which standard error has been reported and convert it 
#https://github.com/GenomicSEM/GenomicSEM/wiki/5.-User-Specified-Models-with-SNP-Effects
#thus I created a function to check if it is SE of logistic beta
is_se_logB <- function(BETA,SE, PVALUE) {
  p_calculated <- 2*pnorm((abs(BETA) / SE),lower.tail = F)
  p_reported <- PVALUE
  data.frame(p_calculated, p_reported)
}
#----END of function -----------------------------------------------------------



#----Load Group 1 sumstats -----------------------------------------------------

uc <- 
