
library(data.table)
library(ChIPpeakAnno)
library(stringr)
library(RColorBrewer)
library(moloc)
library(gprofiler2)
library(GenomicRanges)
library(EnsDb.Hsapiens.v75)
library(biomaRt)
library(ComplexHeatmap)

#------------------ gprofiler --------------------------------------------------


#--------- first find the nearest genes to each lead SNP------------------------
#I used ensmble names, on build 37
factor_loci <- fread('outputs/gwas_uc-cd-psc-ra-sle-t1d-jia-asthma-derma/07_colocalization/moloc_factors/factor_loci_moloc_info.txt', data.table = F) 
# 
# ref_genome <-TxDb.Hsapiens.UCSC.hg19.knownGene
ref_genome <- EnsDb.Hsapiens.v75
ref_genes <- genes(ref_genome)
ref_genes<- ref_genes[ref_genes@elementMetadata$gene_biotype=='protein_coding',]
# 
# ref_genes <- genes(ref_genome, single.strand.genes.only=FALSE )
# ref_genes<- c(unlist(ref_genes))
f_lead <- list()
G_list <- list()

mart <- useDataset("hsapiens_gene_ensembl", useMart(biomart="ENSEMBL_MART_ENSEMBL", host="https://grch37.ensembl.org", path="/biomart/martservice" ,dataset="hsapiens_gene_ensembl")) #select the database to convert and maake sure it is build 37 
factor_loci$closest_gene <- rep('-', nrow(factor_loci))

for(i in 1:3){
  tt <- c('f1', 'f2', 'f3')[i]
  f_lead[[tt]] <- GRanges(seqnames  =factor_loci[factor_loci$trait==tt, c('chr')],   IRanges(names =factor_loci[factor_loci$trait==tt, c('SNP')] , start = factor_loci[factor_loci$trait==tt, 'BP']))
  elementMetadata(f_lead[[tt]])[['ensembl_gene_id']] <-  ref_genes[nearest(f_lead[[tt]], ref_genes, ignore.strand=TRUE, select=c('arbitrary')),]@ranges@NAMES
  
  #add a column with the ensemble gene id into the factor loci table
  factor_loci[factor_loci$trait==tt, ][match(factor_loci[factor_loci$trait==tt, ]$SNP, f_lead[[tt]]@ranges@NAMES),]$closest_gene  <- unlist(elementMetadata(f_lead[[tt]])[['ensembl_gene_id']])
}


fwrite(factor_loci, 'outputs/gwas_uc-cd-psc-ra-sle-t1d-jia-asthma-derma/08_genes_and_pathways/factor_loci_moloc_closest_gene_info.txt') 
factor_loci <- fread('outputs/gwas_uc-cd-psc-ra-sle-t1d-jia-asthma-derma/07_colocalization/factor_loci_moloc_closest_gene_info.txt', data.table = F) 

#plot the upset plot of the genes

q <- make_comb_mat(list(f1=elementMetadata(f_lead$f1)[['ensembl_gene_id']], f2=elementMetadata(f_lead$f2)[['ensembl_gene_id']], f3=elementMetadata(f_lead$f3)[['ensembl_gene_id']]))

UpSet(q)
pdf(width = 10, height = 5, file = 'outputs/gwas_uc-cd-psc-ra-sle-t1d-jia-asthma-derma/08_genes_and_pathways/nearest_genes_upset_plot_factors.pdf')
UpSet(q, set_order = c("f1", "f2", "f3"), comb_order = order(comb_size(q), decreasing = T),
      comb_col = c((brewer.pal(6, 'Set3'))[4:6][comb_degree(q)]),
      top_annotation = upset_top_annotation(q, add_numbers = TRUE, height = unit(6, "cm")),
      right_annotation = upset_right_annotation(q, add_numbers = TRUE, width = unit(5,'cm') )
)
dev.off()

#-------------------------------------------------------------------------------
#write a matrix to be uploaded intro stringr to build the matrix
fwrite(matrix(factor_loci$closest_gene, ncol = 1, nrow = length(factor_loci$closest_gene)), 'outputs/gwas_uc-cd-psc-ra-sle-t1d-jia-asthma-derma/08_genes_and_pathways/test.string.csv')

#read the file generated by string with protein ids and make it readable by biomaRt
string <- fread('outputs/gwas_uc-cd-psc-ra-sle-t1d-jia-asthma-derma/08_genes_and_pathways/STRING_network_node.csv', data.table = F)
string <- string[,c(1,2)]
string[,2] <- str_sub(string$`@id`, start = 15)
string

#find ensembl_gene_id to the proteins from string
s <- getBM(filters= "ensembl_peptide_id", attributes= c("ensembl_peptide_id","ensembl_gene_id"),values=  string[,2] ,mart= mart)
string$ensembl_gene_id <- 
  string$ensembl_gene_id <- s[base::match(string[,2], s$ensembl_peptide_id),]$ensembl_gene_id

#chech that the enemble ids are the ones present in origin in my dataset
mean(string$ensembl_gene_id %in% factor_loci$closest_gene) 


string$trait <- rep(NA, nrow(string))
ttr <- list()
factor_loci_noNA<- factor_loci[!is.na(factor_loci$closest_gene),]
string_no_NA <- string[!is.na(string$ensembl_gene_id),]
string$ensembl_gene_id[is.na(string$ensembl_gene_id)] <- 0

for(i in 1:length(string_no_NA$ensembl_gene_id)) {
  a<- string_no_NA$ensembl_gene_id[i]
  string_no_NA$trait[i]<- paste0(factor_loci_noNA[factor_loci_noNA$closest_gene==a, ]$trait, collapse = '-')
  
}


string$trait <- string_no_NA[match(string$ensembl_gene_id, string_no_NA$ensembl_gene_id),]$trait
string$map <- str_sub(string$`@id`, start = 10) 
fwrite(string, 'outputs/gwas_uc-cd-psc-ra-sle-t1d-jia-asthma-derma/08_genes_and_pathways/string_mapping_columns.csv')




#gprofiler----------------------------------------------------------------------

gost_test_region <- gost( organism = , query = list(f1=unique(elementMetadata(f_lead$f1)[['ensembl_gene_id']]), f2=unique(elementMetadata(f_lead$f2)[['ensembl_gene_id']]), f3=unique(elementMetadata(f_lead$f3)[['ensembl_gene_id']])),
                          sources = c("GO:BP", "REAC", 'KEGG'), significant = T, evcodes = T)
gostplot(gost_test_region, capped = F)
saveRDS(gost_test_region, 'outputs/gwas_uc-cd-psc-ra-sle-t1d-jia-asthma-derma/08_genes_and_pathways/gprofiler_object.RDS')










