---
title: "02_munge_ldsc_sumstats"
author: "Pietro Demela"
date: "21/12/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	message = TRUE,
	warning = TRUE
)
knitr::opts_knit$set(root.dir =  "/project/aid_sharing/AID_sharing") #set up the project directory
```


## LD score regression on auotoimmunity GWAS summary statistics
In this script the selected summary stats will be used to run ldsc, munge and sum stats function. 
Importat: the have to be in the same order in the function

```{r libraries, message=TRUE, warning=TRUE, paged.print=TRUE}
library(data.table)
library(GenomicSEM)
library(tidyr)
library(dplyr)
library(corrplot)
```

## Function for effective sample size calculations
A function for doing it rapidly from csv files. The csv files generated by me, from info in the papers
```{r calculate_prevalence, message=TRUE, warning=TRUE, paged.print=TRUE}

calculate_prevalence <- function(path_of_csv){
  #read the file and remove the NA columns and rows
  preva_csv <- read.csv(file = path_of_csv , header = T, sep = ';')
  preva_csv <-  preva_csv[, c('cases', 'controls')]
  preva_csv_noNA <- preva_csv[complete.cases(preva_csv), ]
  
  #calculated effective prevalence 
  #calculate sample prevalence for each cohort
  preva_csv_noNA$v <-preva_csv_noNA$cases/(preva_csv_noNA$cases+preva_csv_noNA$controls)
  #calculate cohort specific effective sample size
  preva_csv_noNA$EffN<-4*preva_csv_noNA$v*(1-preva_csv_noNA$v)*(preva_csv_noNA$cases+preva_csv_noNA$controls)
  #calculate sum of effective sample size: 
  eff_sample_size <- round(sum(preva_csv_noNA$EffN), 3)
  
  #print number of cases and controls, and effective sample size
  cat(paste0( 'Number of cases  ' , sum(preva_csv_noNA$cases), '\n', 
              'Number of contros  ', sum(preva_csv_noNA$controls), '\n', 
              'Effective sample size  ', round(sum(preva_csv_noNA$EffN), 3)
  ))
  
  #output the prevalenc when assigned to a variable 
  invisible(round(eff_sample_size, 3))
}

```



## Munge

```{r}
file_names <- c('outputs/rev_1/01_prepare/ready_format/t1d_chiou-2021.txt', 
                'outputs/rev_1/01_prepare/ready_format/cd_build37_delange-2017.txt', 
                'outputs/rev_1/01_prepare/ready_format/cd_build37_delange-2017.txt',
                'outputs/rev_1/01_prepare/ready_format/psc_ji-2016.txt',
                'outputs/rev_1/01_prepare/ready_format/jia_beta_lopezisac-2020.txt',
                'outputs/rev_1/01_prepare/ready_format/sle_beta_bentham-2015.txt',
                'outputs//rev_1/01_prepare/ready_format/ra_eu_okada-2014.txt',
                'outputs/rev_1/01_prepare/ready_format/asthma_han-2020.txt',
                'outputs/rev_1/01_prepare/ready_format/derma_sliz-2021.txt'
                )
traits <- c('t1d','crohn', 'uc', 'psc', 'jia', 'sle', 'ra',  'asthma', 'derma')

t1d_p <- calculate_prevalence('Prevalences/CSV_prevalences/t1d_chiou-2021.csv')
crohn_p <- calculate_prevalence('Prevalences/CSV_prevalences/crohn_delange-2017.csv')
uc_p <- calculate_prevalence('Prevalences/CSV_prevalences/uc_delange-2017.csv')
psc_p <- 14890
jia_p <- 12501
sle_p <- calculate_prevalence('Prevalences/CSV_prevalences/sle_benthman-2015.csv')
ra_p <- calculate_prevalence('Prevalences/CSV_prevalences/ra_okada-2014_only-eu.csv')
asthma_p <- 64538 + 329321
derma_p <- calculate_prevalence('Prevalences/CSV_prevalences/derma_sliz-2021.csv')

munge(files = file_names, hm3 = 'SNP/w_hm3.snplist', 
      N = c(t1d_p, crohn_p, uc_p, psc_p, jia_p, sle_p, ra_p, asthma_p, derma_p), 
      trait.names = traits,
      parallel = T,
      cores = 9)
```



